PGM Enumeration

stwrt opened the pull request:
#### e1dfe8ee432e1e56188157ffb1ce6f6afb7758ba
<pre>
PGM Enumeration
<a href="https://bugs.webkit.org/show_bug.cgi?id=252331">https://bugs.webkit.org/show_bug.cgi?id=252331</a>

Reviewed by Brent Fulgham.

Add enumeration support for PGM allocation. This will enumerate both the allocation and
surronding metadata regarding that allocation. Freed PGM allocations are not tracked.

* Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/src/libpas/pas_enumerate_large_heaps.c:
(pas_hash_map_entry_callback):
(enumerate_pgm_map):
(pas_enumerate_large_heaps):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_deallocate):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_root.c:
(pas_root_construct):
* Source/bmalloc/libpas/src/libpas/pas_root.h:
* Source/bmalloc/libpas/src/test/EnumerationTests.cpp: Added.
(std::PageRange::PageRange):
(std::PageRange::operator&lt; const):
(std::PageRange::end const):
(std::RecordedRange::RecordedRange):
(std::RecordedRange::operator&lt; const):
(std::RecordedRange::end const):
(std::ReaderRange::ReaderRange):
(std::ReaderRange::operator&lt; const):
(std::enumeratorReader):
(std::enumeratorRecorder):
(std::testBasicEnumeration):
(std::testPGMEnumerationBasic):
(std::testPGMEnumerationAddAndFree):
(addEnumerationTests):
* Source/bmalloc/libpas/src/test/TestHarness.cpp:
(main):

Canonical link: <a href="https://commits.webkit.org/260978@main">https://commits.webkit.org/260978@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/01647f1fe6225bf97451bf198f74abbfac831810

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110137 "22 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19236 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1565 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20697 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10426 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102393 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115883 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/30279 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/98915 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11931 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31616 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/99874 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9972 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/12545 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31202 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17910 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51232 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/107895 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7608 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14349 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26613 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -86,12 +87,12 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
+          * *------------------* *------------------* *------------------*
+          * | lower guard page | | user alloc pages | | upper guard page |
+          * *------------------* *------------------* *------------------*
+          */
```
This whitespace change seems wrong.

brentfulgham commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -111,9 +112,9 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
-     */
+           * ensure physical addresses are released
+           * TODO: investigate using MADV_FREE_REUSABLE instead
+           */
```
Ditto: This whitespace/indentation change seems wrong.

stwrt replied:
Fixed

brentfulgham commented on Source/bmalloc/libpas/src/test/EnumerationTests.cpp:
```diff
@@ -0,0 +1,315 @@
+#include <map>
+#include <stdlib.h>
+#include <set>
+#include <sys/mman.h>
```
Nit: usually we would have these `stdlib` includes appear after the `libpas` headers. Does this code not build if they appear in our usual order?

stwrt replied:
Builds fine.

brentfulgham commented on Source/bmalloc/libpas/src/test/EnumerationTests.cpp:
```diff
@@ -0,0 +1,315 @@
+void* enumeratorReader(pas_enumerator* enumerator,
+                       void* address,
+                       size_t size,
+                       void* arg)
```
Nit: We don't generally declare function arguments with each on their own line.

stwrt replied:
This is code taken from another test case that Phil wrote.

I changed it to one line.

brentfulgham commented on Source/bmalloc/libpas/src/test/EnumerationTests.cpp:
```diff
@@ -0,0 +1,315 @@
+        pas_round_up_to_power_of_2(
+            reinterpret_cast<uintptr_t>(address) + size,
+            pas_page_malloc_alignment())
+        - reinterpret_cast<uintptr_t>(pageAddress);
```
Ditto: These definitions are not being done in standard WebKit style. Does `libpas` have its own style?

stwrt replied:
Yes, it does have some different style choices... but I don't think it would be bad to more closely align with WebKit at this point.

brentfulgham commented on Source/bmalloc/libpas/src/test/EnumerationTests.cpp:
```diff
@@ -0,0 +1,315 @@
+    if (verbose)
+        cout << "Testing basic enuemration" << endl;
+
+    /* create root */
```
Nit: The "create root" comment doesn't really add much to the function call for `pas_root_create`

stwrt replied:
Yeah, I removed these comments. They were more just some extra comments remaining around from development.

brentfulgham reviewed (approved):
The changes look fine, but @Constellation could probably provide better technical comments. The code doesn't seem to follow general WebKit style, so I'd suggest you address that unless PGM uses its own style guide.

stwrt committed:
PGM Enumeration
https://bugs.webkit.org/show_bug.cgi?id=252331

Reviewed by Brent Fulgham.

Add enumeration support for PGM allocation. This will enumerate both the allocation and
surronding metadata regarding that allocation. Freed PGM allocations are not tracked.

* Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/src/libpas/pas_enumerate_large_heaps.c:
(pas_hash_map_entry_callback):
(enumerate_pgm_map):
(pas_enumerate_large_heaps):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_deallocate):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_root.c:
(pas_root_construct):
* Source/bmalloc/libpas/src/libpas/pas_root.h:
* Source/bmalloc/libpas/src/test/EnumerationTests.cpp: Added.
(std::PageRange::PageRange):
(std::PageRange::operator< const):
(std::PageRange::end const):
(std::RecordedRange::RecordedRange):
(std::RecordedRange::operator< const):
(std::RecordedRange::end const):
(std::ReaderRange::ReaderRange):
(std::ReaderRange::operator< const):
(std::enumeratorReader):
(std::enumeratorRecorder):
(std::testBasicEnumeration):
(std::testPGMEnumerationBasic):
(std::testPGMEnumerationAddAndFree):
(addEnumerationTests):
* Source/bmalloc/libpas/src/test/TestHarness.cpp:
(main):

Canonical link: https://commits.webkit.org/260978@main
